src/cpu/aarch64/matmul/acl_matmul.hpp

```diff
@@ -96,12 +96,24 @@ struct acl_matmul_t : public primitive_t {
             acl_obj_->src_acc_tensor.allocator()->init(amp.src_acc_info);
             acl_obj_->transA.configure(
                     &acl_obj_->src_acc_tensor, &acl_obj_->src_tensor);
+
+            acl_obj_->src_intermediate_tensor.allocator()->init(amp.src_info);
+            acl_obj_->src_intermediate_tensor.allocator()->allocate();
```

Comment:
Unfortunately, this is contrary to oneDNN memory model. oneDNN typically does not hog memory inside primitive objects.
oneDNN calls these temporary memory "scratchpad memory", and it is either allocated during execution (library mode) or provided by user (user mode). See [this page ](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_scratchpad.html#primitive-attributes-scratchpad) for more details.

Reply:
Hi @mgouicem, thanks for the feedback. I have some high level question on the onednn primitive caching feature, could you please clarify:
if we expect the primitive caching to not to hold on to any primitive memory, then what exactly the caching feature is saving in the execution time? In the entire primitive create/configure/execute states I see tensor memory allocation is the major time consuming operation.  Given tensors could be of > 100MB for some CNN networks, allocating the buffers during execute, every time, may not be acceptable. I consider these intermediate tensors as 'workspace' buffers as these are not known to user to be able to provide them via scratchpad.

Reply:
Hi @snadampal , the primitive cache mainly reduces overheads of creating primitive objects. For the acl implementations, it might not be much, but for other implementations (e.g. sve, and gpu), this overhead of generating executable binaries taylored to a given problem shapes can be large.

Regarding memory allocation, oneDNN is typically called within a larger FWK (e.g. tensorflow, pytorch, OpenVINO, ...). Those have dedicated memory allocator that allow to get memory pages from the OS, and just reuse them across multiple calls of oneDNN. To do that, they create primitives using the scratchpad user_mode, and pass the scratchpad memory directly to the primitive execution function. Hence they manage the lifetime of the scratchpad as they wish.

If we were to keep scratchpad memory alive inside the oneDNN primitive cache, this could lead to large memory consumption from oneDNN, and this is not acceptable on some devices (e.g. client devices, GPUs with limited memory, ...).

Hope this helps clarify this design choice.

Reply:
@mgouicem really appreciate your time for putting all these details, I got the rational for the current design, thank you!
Now I understand why you have primitive and resource objects separate and why the resources are not cached. Yes, both TensorFlow and PyTorch are using the scratchpad user_mode, but for aarch64 I see ACL managing the memory allocations and these tensors are being imported everywhere else. I will revisit the tensor buffer management flow again.